template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -291,6 +311,52 @@ public String getStackVersion() {
         return cmTemplate.getCdhVersion();
     }
 
+    @Override
+    public Map<String, Map<String, ServiceAttributes>> getHostGroupBasedServiceAttributes() {
```

Comment:
I ended up with this, but the method and variable names should be changed with more meaningful one I think. This is just an example:
```    
	@Override
    public Map<String, Map<String, ServiceAttributes>> getHostGroupBasedServiceAttributes() {
        Map<String, Set<String>> componentsByHostGroup = collectComponentsByHostrgoup();
        Set<String> computeHostGroups = getRecommendationByBlacklist(BlackListedLoadBasedAutoscaleRole.class,
                true, componentsByHostGroup);
        Map<String, Map<String, ServiceAttributes>> result = createServiceAttributeMap(componentsByHostGroup, computeHostGroups);
        LOGGER.debug("ServiceAttributes: {}", result);
        return result;
    }

    private Map<String, Map<String, ServiceAttributes>> createServiceAttributeMap(Map<String, Set<String>> componentsByHostGroup, Set<String> computeHostGroups) {
        Map<String, Map<String, ServiceAttributes>> result = new HashMap<>();
        for (String hg : componentsByHostGroup.keySet()) {
            String instanceType = computeHostGroups.contains(hg) ? YarnConstants.ATTRIBUTE_NODE_INSTANCE_TYPE_COMPUTE
                    : YarnConstants.ATTRIBUTE_NODE_INSTANCE_TYPE_WORKER;
            Map<String, ServiceAttributes> m = Collections.singletonMap(YarnRoles.YARN, new ServiceAttributes(ServiceComponent.of(YarnRoles.YARN, YarnRoles.NODEMANAGER),
                    Collections.singletonMap(YarnConstants.ATTRIBUTE_NAME_NODE_INSTANCE_TYPE, instanceType)));
            result.put(hg, m);
        }
        return result;
    }

    private Map<String, Set<String>> collectComponentsByHostrgoup() {
        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponentsWithYarnNMs = hgToNonGwServiceComponents.entrySet().stream()
                .filter(this::isYarnNodemanager)
                .collect(toMap(Entry::getKey, Entry::getValue));
        Map<String, Set<String>> componentsByHostGroup = hgToNonGwServiceComponentsWithYarnNMs.entrySet()
                .stream().collect(toMap(Entry::getKey, this::collectComponents));
        return componentsByHostGroup;
    }

    private Set<String> collectComponents(Entry<String, Set<ServiceComponent>> e) {
        return e.getValue().stream()
                .map(ServiceComponent::getComponent)
                .collect(Collectors.toUnmodifiableSet());
    }

    private boolean isYarnNodemanager(Entry<String, Set<ServiceComponent>> entry) {
        return entry.getValue().stream().anyMatch(sc -> YarnRoles.YARN.equalsIgnoreCase(sc.getService())
                && YarnRoles.NODEMANAGER.equalsIgnoreCase(sc.getComponent()));
    }

Reply:
Thanks for the suggestions. Have broken up the method a bit, and made some other changes.